File: Task_4/code/python/benchmarks/performance_benchmark.py
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'production'))

import numpy as np
import time
import tracemalloc
import psutil
import json
from typing import Dict, List
from matrix_multiplier import DistributedMatrixMultiplier, basic_multiply, parallel_multiply


class PerformanceBenchmark:
    """
    Benchmark suite for comparing different matrix multiplication approaches.
    Measures execution time, memory usage, and scalability.
    """
    
    def __init__(self):
        self.results = []
        self.multiplier = None  # Only initialize when needed for distributed
        
    def _measure_performance(self, func, *args, num_runs: int = 5, is_distributed: bool = False) -> Dict:
        """Measure execution time and memory usage of a function with multiple runs."""
        execution_times = []
        memory_used = []
        network_times = []
        data_transferred = []
        
        # Warmup runs (3 iterations)
        import gc
        print("  Warmup: ", end="", flush=True)
        for w in range(3):
            _ = func(*args)
            gc.collect()
            print(f".{w+1}", end="", flush=True)
        print(" done")
        time.sleep(0.1)
        
        for run in range(num_runs):
            # Force garbage collection for consistent memory measurements
            import gc
            gc.collect()
            
            # Start memory tracking
            tracemalloc.start()
            
            # Measure execution time
            start_time = time.time()
            result = func(*args)
            end_time = time.time()
            
            # Get peak memory usage
            current, peak = tracemalloc.get_traced_memory()
            tracemalloc.stop()
            
            exec_time_ms = (end_time - start_time) * 1000
            mem_kb = peak / 1024
            
            execution_times.append(exec_time_ms)
            memory_used.append(mem_kb)
            
            # Print progress
            print(f"    Run {run + 1}/{num_runs} completed: {exec_time_ms:.2f}ms, {mem_kb:.2f} KB")
            
            # Collect distributed-specific metrics
            if is_distributed and hasattr(self.multiplier, 'network_time'):
                network_times.append(self.multiplier.network_time * 1000)  # Convert to ms
                data_transferred.append(self.multiplier.data_transferred_mb)
        
        metrics = {
            'execution_times': execution_times,
            'avg_execution_time': sum(execution_times) / len(execution_times),
            'min_execution_time': min(execution_times),
            'max_execution_time': max(execution_times),
            'memory_used_kbs': memory_used,  # Per-run memory values
            'memory_used_kb': sum(memory_used) / len(memory_used),  # Average for display
            'num_runs': num_runs,
            'result_shape': result.shape if hasattr(result, 'shape') else None
        }
        
        # Add distributed-specific metrics (both per-run and averages)
        if network_times:
            metrics['network_times'] = network_times  # Per-run network times
            metrics['data_transferred_mbs'] = data_transferred  # Per-run data transfer
            metrics['network_time_avg'] = sum(network_times) / len(network_times)
            metrics['data_transferred_mb_avg'] = sum(data_transferred) / len(data_transferred)
            metrics['computation_time_avg'] = metrics['avg_execution_time'] - metrics['network_time_avg']
            metrics['network_overhead_percent'] = (metrics['network_time_avg'] / metrics['avg_execution_time']) * 100
        
        return metrics
    
    def benchmark_matrix_sizes(self, sizes: List[int], distributed_only: bool = False, method: str = 'all') -> List[Dict]:
        """Benchmark different matrix sizes.
        
        Args:
            sizes: List of matrix sizes to test
            distributed_only: If True, only run distributed tests (deprecated, use method='distributed')
            method: Which method to run ('all', 'basic', 'parallel', 'distributed')
        """
        print("Starting matrix size benchmark...")
        
        # Only connect to Hazelcast if we need distributed tests
        distributed_available = False
        if method in ['all', 'distributed']:
            self.multiplier = DistributedMatrixMultiplier()
            if not self.multiplier.connect():
                print("Warning: Could not connect to Hazelcast. Skipping distributed tests.")
                distributed_available = False
            else:
                distributed_available = True
        
        results = []
        
        for size in sizes:
            print(f"\nTesting matrix size: {size}x{size}")
            
            # Generate test matrices
            A = np.random.rand(size, size)
            B = np.random.rand(size, size)
            
            num_runs = 5
            
            test_result = {
                'matrix_size': size,
                'total_elements': size * size * 2,
                'memory_size_mb': (A.nbytes + B.nbytes) / 1024 / 1024,
                'num_runs': num_runs
            }
            
            if not distributed_only and method in ['all', 'basic']:
                # Test basic multiplication
                print(f"  Testing basic multiplication ({num_runs} runs)...")
                basic_perf = self._measure_performance(basic_multiply, A, B, num_runs=num_runs)
                test_result['basic'] = basic_perf
                
            if not distributed_only and method in ['all', 'parallel']:
                # Test parallel multiplication
                print(f"  Testing parallel multiplication ({num_runs} runs)...")
                parallel_perf = self._measure_performance(parallel_multiply, A, B, num_runs=num_runs)
                test_result['parallel'] = parallel_perf
            
            # Test distributed multiplication (if available and requested)
            if distributed_available and method in ['all', 'distributed']:
                print(f"  Testing distributed multiplication ({num_runs} runs)...")
                try:
                    dist_perf = self._measure_performance(self.multiplier.multiply, A, B, num_runs=num_runs, is_distributed=True)
                    dist_perf['cluster_nodes'] = self.multiplier.cluster_size
                    dist_perf['memory_per_node_kb'] = self.multiplier.memory_per_node_mb * 1024  # Convert MB to KB
                    test_result['distributed'] = dist_perf
                except Exception as e:
                    print(f"    Distributed test failed: {e}")
                    test_result['distributed'] = {'error': str(e)}
            elif method == 'distributed':
                test_result['distributed'] = {'error': 'Hazelcast not available'}
            
            results.append(test_result)
            
            # Print summary for this size
            self._print_size_summary(test_result)
        
        if distributed_available and self.multiplier:
            self.multiplier.disconnect()
        
        return results
    
    def _print_size_summary(self, result: Dict):
        """Print summary for a single matrix size test."""
        size = result['matrix_size']
        print(f"\n  Results for {size}x{size} matrices (average of {result.get('num_runs', 5)} runs):")
        
        if 'basic' in result:
            avg_time = result['basic'].get('avg_execution_time', result['basic'].get('execution_time', 0))
            mem = result['basic'].get('memory_used_kb', 0)
            print(f"    Basic:       {avg_time:.2f}ms (min: {result['basic'].get('min_execution_time', avg_time):.2f}ms, max: {result['basic'].get('max_execution_time', avg_time):.2f}ms, Mem: {mem:.2f} KB)")
        if 'parallel' in result:
            avg_time = result['parallel'].get('avg_execution_time', result['parallel'].get('execution_time', 0))
            mem = result['parallel'].get('memory_used_kb', 0)
            print(f"    Parallel:    {avg_time:.2f}ms (min: {result['parallel'].get('min_execution_time', avg_time):.2f}ms, max: {result['parallel'].get('max_execution_time', avg_time):.2f}ms, Mem: {mem:.2f} KB)")
        if 'distributed' in result:
            if 'avg_execution_time' in result['distributed'] or 'execution_time' in result['distributed']:
                avg_time = result['distributed'].get('avg_execution_time', result['distributed'].get('execution_time', 0))
                network_time = result['distributed'].get('network_time_avg', 0)
                network_pct = result['distributed'].get('network_overhead_percent', 0)
                data_mb = result['distributed'].get('data_transferred_mb_avg', 0)
                nodes = result['distributed'].get('cluster_nodes', 1)
                mem_kb = result['distributed'].get('memory_used_kb', 0)
                print(f"    Distributed: {avg_time:.2f}ms (min: {result['distributed'].get('min_execution_time', avg_time):.2f}ms, max: {result['distributed'].get('max_execution_time', avg_time):.2f}ms)")
                print(f"                 Network: {network_time:.2f}ms ({network_pct:.1f}% overhead), Data: {data_mb:.2f} MB transferred")
                print(f"                 Cluster: {nodes} node(s), Memory: {mem_kb:.2f} KB)")
            else:
                print(f"    Distributed: Failed ({result['distributed']['error']})")
    
    def run_scalability_test(self, sizes=None, distributed_only: bool = False, method: str = 'all') -> Dict:
        """Run comprehensive scalability test."""
        if sizes is None:
            sizes = [64, 128, 256]
        
        print("=" * 60)
        print("SCALABILITY BENCHMARK")
        print("=" * 60)
        
        # Test different matrix sizes
        results = self.benchmark_matrix_sizes(sizes, distributed_only=distributed_only, method=method)
        
        # Calculate scalability metrics
        scalability_analysis = self._analyze_scalability(results)
        
        return {
            'benchmark_results': results,
            'scalability_analysis': scalability_analysis,
            'timestamp': time.strftime('%Y-%m-%d %H:%M:%S'),
            'system_info': self._get_system_info()
        }
    
    def _analyze_scalability(self, results: List[Dict]) -> Dict:
        """Analyze scalability from benchmark results."""
        analysis = {}
        
        if len(results) < 2:
            return analysis
        
        # Analyze each method
        for method in ['basic', 'parallel', 'distributed']:
            method_times = []
            method_sizes = []
            
            for result in results:
                if method in result:
                    # Use avg_execution_time if available, otherwise execution_time
                    exec_time = result[method].get('avg_execution_time', result[method].get('execution_time'))
                    if exec_time is not None:
                        method_times.append(exec_time)
                        method_sizes.append(result['matrix_size'])
            
            if len(method_times) >= 2:
                # Calculate speedup relative to smallest size
                base_time = method_times[0]
                speedups = [base_time / t if t > 0 else 0 for t in method_times]
                
                # Calculate efficiency (theoretical vs actual)
                size_ratios = [s / method_sizes[0] for s in method_sizes]
                theoretical_ratios = [r ** 3 for r in size_ratios]  # O(n³) complexity
                
                analysis[method] = {
                    'execution_times': method_times,
                    'matrix_sizes': method_sizes,
                    'speedups': speedups,
                    'size_ratios': size_ratios,
                    'theoretical_complexity_ratios': theoretical_ratios
                }
        
        return analysis
    
    def _get_system_info(self) -> Dict:
        """Get system information for benchmark context."""
        return {
            'cpu_count': psutil.cpu_count(),
            'cpu_count_logical': psutil.cpu_count(logical=True),
            'total_memory_gb': psutil.virtual_memory().total / 1024 / 1024 / 1024,
            'available_memory_gb': psutil.virtual_memory().available / 1024 / 1024 / 1024,
            'python_version': sys.version.split()[0]
        }
    
    def save_results(self, results: Dict, filename: str = None, method_suffix: str = None):
        """Save benchmark results to CSV file (matching Java format)."""
        if filename is None:
            timestamp = int(time.time() * 1000)  # milliseconds since epoch
            
            # Get cluster size from distributed results if available
            cluster_size = 0
            if method_suffix == 'distributed' and results.get('benchmark_results'):
                for result in results['benchmark_results']:
                    if 'distributed' in result and 'cluster_nodes' in result['distributed']:
                        cluster_size = result['distributed']['cluster_nodes']
                        break
            
            if method_suffix:
                if method_suffix == 'distributed' and cluster_size > 0:
                    filename = f'benchmark_results_python_{method_suffix}_{cluster_size}nodes_{timestamp}.csv'
                else:
                    filename = f'benchmark_results_python_{method_suffix}_{timestamp}.csv'
            else:
                filename = f'benchmark_results_python_{timestamp}.csv'
        
        # Save to results directory
        results_dir = os.path.join(os.path.dirname(__file__), '..', 'results')
        os.makedirs(results_dir, exist_ok=True)
        filepath = os.path.join(results_dir, filename)
        
        # Write CSV in same format as Java
        with open(filepath, 'w') as f:
            f.write('matrix_size,method,run,time_ms,network_time_ms,data_transferred_kb,memory_kb,cluster_nodes,memory_per_node_kb\n')
            
            for result in results.get('benchmark_results', []):
                size = result['matrix_size']
                
                # Write basic results
                if 'basic' in result:
                    basic = result['basic']
                    memory_kbs = basic.get('memory_used_kbs', [])
                    for i, time_ms in enumerate(basic['execution_times'], 1):
                        mem_kb = memory_kbs[i-1] if i-1 < len(memory_kbs) else basic.get('memory_used_kb', 0)
                        f.write(f'{size},basic,{i},{time_ms:.2f},0,0,{mem_kb:.2f},0,0\n')
                
                # Write parallel results
                if 'parallel' in result:
                    parallel = result['parallel']
                    memory_kbs = parallel.get('memory_used_kbs', [])
                    for i, time_ms in enumerate(parallel['execution_times'], 1):
                        mem_kb = memory_kbs[i-1] if i-1 < len(memory_kbs) else parallel.get('memory_used_kb', 0)
                        f.write(f'{size},parallel,{i},{time_ms:.2f},0,0,{mem_kb:.2f},0,0\n')
                
                # Write distributed results
                if 'distributed' in result and 'avg_execution_time' in result['distributed']:
                    dist = result['distributed']
                    cluster_nodes = dist.get('cluster_nodes', 1)
                    memory_per_node_kb = dist.get('memory_per_node_kb', 0)
                    network_times = dist.get('network_times', [])
                    data_mbs = dist.get('data_transferred_mbs', [])
                    memory_kbs = dist.get('memory_used_kbs', [])
                    
                    for i, time_ms in enumerate(dist['execution_times'], 1):
                        # Use per-run values if available, otherwise use averages
                        network_ms = network_times[i-1] if i-1 < len(network_times) else dist.get('network_time_avg', 0)
                        data_mb = data_mbs[i-1] if i-1 < len(data_mbs) else dist.get('data_transferred_mb_avg', 0)
                        data_kb = data_mb * 1024  # Convert MB to KB
                        mem_kb = memory_kbs[i-1] if i-1 < len(memory_kbs) else dist.get('memory_used_kb', 0)
                        f.write(f'{size},distributed,{i},{time_ms:.2f},{network_ms:.2f},{data_kb:.2f},{mem_kb:.2f},{cluster_nodes},{memory_per_node_kb:.2f}\n')
        
        print(f"\nResults saved to: {filepath}")
        return filepath


def main():
    """Main benchmark execution."""
    import argparse
    
    parser = argparse.ArgumentParser(description='Matrix Multiplication Benchmark')
    parser.add_argument('--sizes', type=str, default='16,32,64,128',
                        help='Comma-separated matrix sizes to test (e.g., 16,32,64,128)')
    parser.add_argument('--method', type=str, choices=['basic', 'parallel', 'distributed', 'all'], 
                        default='all',
                        help='Which method to benchmark (basic, parallel, distributed, or all)')
    args = parser.parse_args()
    
    # Parse sizes
    sizes = [int(s.strip()) for s in args.sizes.split(',')]
    
    # Determine which methods to run
    if args.method == 'all':
        distributed_only = False
        method_suffix = None
    elif args.method == 'distributed':
        distributed_only = True
        method_suffix = 'distributed'
    elif args.method in ['basic', 'parallel']:
        # For basic/parallel only, we'll still need to run both but filter in save
        distributed_only = False
        method_suffix = args.method
    else:
        distributed_only = False
        method_suffix = None
    
    benchmark = PerformanceBenchmark()
    
    print("Matrix Multiplication Benchmark")
    print("=" * 50)
    print(f"Mode: {args.method.capitalize()} method{'s' if args.method == 'all' else ''}")
    
    # Run scalability test with custom sizes
    results = benchmark.run_scalability_test(sizes, distributed_only=(args.method == 'distributed'), method=args.method)
    
    # Save results
    filepath = benchmark.save_results(results, method_suffix=method_suffix)
    
    # Print final summary
    print("\n" + "=" * 60)
    print("BENCHMARK SUMMARY")
    print("=" * 60)
    
    if 'scalability_analysis' in results:
        analysis = results['scalability_analysis']
        
        print("\nExecution time trends:")
        for method, data in analysis.items():
            if 'execution_times' in data:
                times = data['execution_times']
                sizes = data['matrix_sizes']
                print(f"  {method.capitalize()}:")
                for size, exec_time in zip(sizes, times):
                    print(f"    {size}x{size}: {exec_time:.2f}ms")
    
    print(f"\nDetailed results saved to: {filepath}")
    print("\nRecommendations:")
    print("- For small matrices (< 200x200): Use parallel (NumPy)")
    print("- For large matrices (> 1000x1000): Use distributed (Hazelcast)")
    print("- Network overhead is significant for small matrices")
    print("- Distributed approach scales better with matrix size")


if __name__ == "__main__":
    main()

File: Task_4/code/java/src/main/java/com/ulpgc/bigdata/benchmarks/PerformanceBenchmark.java
package com.ulpgc.bigdata.benchmarks;

import com.ulpgc.bigdata.production.DistributedMatrixMultiplier;

import java.io.FileWriter;
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Performance benchmark for matrix multiplication approaches.
 */
public class PerformanceBenchmark {
    
    private static final int NUM_RUNS = 5;
    private static final int[] MATRIX_SIZES = {64, 128, 256, 512, 1024, 2048};
    private static String METHOD = "all";  // all, basic, parallel, distributed
    
    public static void main(String[] args) {
        // Parse command line arguments
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--method") && i + 1 < args.length) {
                METHOD = args[i + 1];
            }
        }
        
        System.out.println("============================================================");
        System.out.println("MATRIX MULTIPLICATION PERFORMANCE BENCHMARK (Java)");
        System.out.println("============================================================");
        System.out.println("Configuration:");
        System.out.println("  - Runs per test: " + NUM_RUNS);
        System.out.println("  - Matrix sizes: " + arrayToString(MATRIX_SIZES));
        System.out.println("  - CPU cores: " + Runtime.getRuntime().availableProcessors());
        System.out.println("  - Mode: " + METHOD.substring(0, 1).toUpperCase() + METHOD.substring(1) + 
                          (METHOD.equals("all") ? " methods" : " method"));
        System.out.println("============================================================\n");
        
        DistributedMatrixMultiplier multiplier = new DistributedMatrixMultiplier();
        boolean distributedAvailable = false;
        
        // Only connect to Hazelcast if distributed method is needed
        if (METHOD.equals("all") || METHOD.equals("distributed")) {
            distributedAvailable = multiplier.connect();
            
            if (!distributedAvailable) {
                System.out.println("WARNING: Could not connect to Hazelcast. Distributed tests will be skipped.\n");
            }
        }
        
        List<BenchmarkResult> allResults = new ArrayList<>();
        
        try {
            for (int size : MATRIX_SIZES) {
                System.out.println("Testing matrix size: " + size + "x" + size);
                System.out.println("------------------------------------------------------------");
                
                BenchmarkResult result = null;
                try {
                    result = benchmarkSize(size, multiplier, distributedAvailable);
                    allResults.add(result);
                    printSizeSummary(result);
                } catch (InterruptedException e) {
                    System.err.println("Benchmark interrupted: " + e.getMessage());
                    Thread.currentThread().interrupt();
                }
                System.out.println();
            }
            
            // Print overall summary
            printOverallSummary(allResults);
            
            // Save results to file with method suffix
            String methodSuffix = METHOD.equals("all") ? null : METHOD;
            int clusterSize = distributedAvailable ? multiplier.getClusterSize() : 0;
            saveResults(allResults, methodSuffix, clusterSize);
            
        } finally {
            if (distributedAvailable) {
                multiplier.disconnect();
            }
        }
    }
    
    private static BenchmarkResult benchmarkSize(int size, DistributedMatrixMultiplier multiplier, boolean distributedAvailable) throws InterruptedException {
        BenchmarkResult result = new BenchmarkResult(size);
        Runtime runtime = Runtime.getRuntime();
        
        // Generate test matrices
        double[][] A = DistributedMatrixMultiplier.generateRandomMatrix(size, size);
        double[][] B = DistributedMatrixMultiplier.generateRandomMatrix(size, size);
        
        if (METHOD.equals("all") || METHOD.equals("basic")) {
            // Basic multiplication
            System.out.println("  Testing basic multiplication (" + NUM_RUNS + " runs)...");
            // Warmup runs (3 iterations for JIT compilation)
            System.out.print("  Warmup: ");
            for (int w = 0; w < 3; w++) {
                DistributedMatrixMultiplier.basicMultiply(A, B);
                System.gc();
                System.out.print("." + (w + 1));
            }
            System.out.println(" done");
            Thread.sleep(100);
            
            result.basicTimes = new double[NUM_RUNS];
            result.memoryUsed = new double[NUM_RUNS];
            for (int i = 0; i < NUM_RUNS; i++) {
                System.gc();
                System.gc();
                Thread.sleep(100);
                long memBefore = runtime.totalMemory() - runtime.freeMemory();
                long start = System.nanoTime();
                double[][] resultMatrix = DistributedMatrixMultiplier.basicMultiply(A, B);
                long memAfter = runtime.totalMemory() - runtime.freeMemory();
                long end = System.nanoTime();
                result.basicTimes[i] = (end - start) / 1_000_000.0;
                result.memoryUsed[i] = Math.max(0.0, (memAfter - memBefore) / 1024.0);
                // Keep reference to prevent GC
                if (resultMatrix[0][0] < -999999) System.out.println("");
            }
        } else {
            result.basicTimes = new double[0];
            result.memoryUsed = new double[0];
        }
        
        if (METHOD.equals("all") || METHOD.equals("parallel")) {
            // Parallel multiplication
            System.out.println("  Testing parallel multiplication (" + NUM_RUNS + " runs)...");
            // Warmup runs (3 iterations for JIT compilation)
            System.out.print("  Warmup: ");
            for (int w = 0; w < 3; w++) {
                DistributedMatrixMultiplier.parallelMultiply(A, B);
                System.gc();
                System.out.print("." + (w + 1));
            }
            System.out.println(" done");
            Thread.sleep(100);
            
            result.parallelTimes = new double[NUM_RUNS];
            result.parallelMemory = new double[NUM_RUNS];
            for (int i = 0; i < NUM_RUNS; i++) {
                System.gc();
                System.gc();
                Thread.sleep(100);
                long memBefore = runtime.totalMemory() - runtime.freeMemory();
                long start = System.nanoTime();
                double[][] resultMatrix = DistributedMatrixMultiplier.parallelMultiply(A, B);
                long memAfter = runtime.totalMemory() - runtime.freeMemory();
                long end = System.nanoTime();
                result.parallelTimes[i] = (end - start) / 1_000_000.0;
                result.parallelMemory[i] = Math.max(0.0, (memAfter - memBefore) / 1024.0);
                // Keep reference to prevent GC
                if (resultMatrix[0][0] < -999999) System.out.println("");
            }
        } else {
            result.parallelTimes = new double[0];
            result.parallelMemory = new double[0];
        }
        
        // Distributed multiplication
        if (distributedAvailable) {
            System.out.println("  Testing distributed multiplication (" + NUM_RUNS + " runs)...");
            // Warmup runs (3 iterations for JIT compilation)
            System.out.print("  Warmup: ");
            for (int w = 0; w < 3; w++) {
                multiplier.multiply(A, B);
                System.gc();
                System.out.print("." + (w + 1));
            }
            System.out.println(" done");
            Thread.sleep(100);
            
            result.distributedTimes = new double[NUM_RUNS];
            result.networkTimes = new double[NUM_RUNS];
            result.dataTransferred = new double[NUM_RUNS];
            result.distributedMemory = new double[NUM_RUNS];
            result.clusterNodes = multiplier.getClusterSize();
            result.memoryPerNodeKB = 0.0;
            
            for (int i = 0; i < NUM_RUNS; i++) {
                System.gc();
                System.gc();
                Thread.sleep(100);
                long memBefore = runtime.totalMemory() - runtime.freeMemory();
                long start = System.nanoTime();
                double[][] resultMatrix = multiplier.multiply(A, B);
                long memAfter = runtime.totalMemory() - runtime.freeMemory();
                long end = System.nanoTime();
                result.distributedTimes[i] = (end - start) / 1_000_000.0;
                result.networkTimes[i] = multiplier.getNetworkTimeMs();
                result.dataTransferred[i] = multiplier.getDataTransferredMB();
                result.distributedMemory[i] = Math.max(0.0, (memAfter - memBefore) / 1024.0);
                // Keep reference to prevent GC
                if (resultMatrix[0][0] < -999999) System.out.println("");
            }
            
            result.clusterNodes = multiplier.getClusterSize();
            result.memoryPerNodeKB = multiplier.getMemoryPerNodeMB() * 1024; // Convert MB to KB
        }
        
        return result;
    }
    
    private static long[] measurePerformance(Runnable task, int numRuns) {
        long[] times = new long[numRuns];
        
        for (int i = 0; i < numRuns; i++) {
            long start = System.currentTimeMillis();
            task.run();
            long end = System.currentTimeMillis();
            times[i] = end - start;
        }
        
        return times;
    }
    
    private static void printSizeSummary(BenchmarkResult result) {
        System.out.println("\n  Results for " + result.size + "x" + result.size + " matrices:");
        System.out.println("    Basic:       " + formatTime(result.basicTimes) + 
                          " | Memory: " + String.format("%.2f KB", averageDouble(result.memoryUsed)));
        System.out.println("    Parallel:    " + formatTime(result.parallelTimes) + 
                          " | Memory: " + String.format("%.2f KB", averageDouble(result.parallelMemory)));
        
        if (result.distributedTimes != null) {
            double avgNetworkTime = averageDouble(result.networkTimes);
            double avgExecTime = averageDouble(result.distributedTimes);
            double networkPercent = (avgNetworkTime / avgExecTime) * 100;
            double avgDataTransferred = averageDouble(result.dataTransferred);
            double avgMemory = averageDouble(result.distributedMemory);
            
            System.out.println("    Distributed: " + formatTime(result.distributedTimes));
            System.out.println("                 Network: " + String.format("%.2fms (%.1f%% overhead) | Data: %.2f MB transferred", 
                              avgNetworkTime, networkPercent, avgDataTransferred));
            System.out.println("                 Cluster: " + result.clusterNodes + " node(s) | Memory: " + String.format("%.2f KB", avgMemory));
        } else {
            System.out.println("    Distributed: Not available");
        }
    }
    
    private static String formatTime(double[] times) {
        if (times == null || times.length == 0) {
            return "N/A";
        }
        
        double avg = averageDouble(times);
        double min = minDouble(times);
        double max = maxDouble(times);
        
        return String.format("%.2fms (min: %.2fms, max: %.2fms)", 
                           avg, min, max);
    }
    
    private static void printOverallSummary(List<BenchmarkResult> results) {
        System.out.println("============================================================");
        System.out.println("OVERALL SUMMARY");
        System.out.println("============================================================");
        System.out.println();
        
        System.out.printf("%-12s %-15s %-15s %-15s%n", "Matrix Size", "Basic (ms)", "Parallel (ms)", "Distributed (ms)");
        System.out.println("------------------------------------------------------------");
        
        for (BenchmarkResult result : results) {
            System.out.printf("%-12s %-15s %-15s %-15s%n",
                result.size + "x" + result.size,
                String.format("%.2fms", averageDouble(result.basicTimes)),
                String.format("%.2fms", averageDouble(result.parallelTimes)),
                result.distributedTimes != null ? 
                    String.format("%.2fms", averageDouble(result.distributedTimes)) : "N/A"
            );
        }
        
        System.out.println("============================================================");
    }
    
    private static void saveResults(List<BenchmarkResult> results, String methodSuffix, int clusterSize) {
        // Create results directory if it doesn't exist (relative to working directory)
        String resultsDir = System.getProperty("user.dir") + "/results";
        new java.io.File(resultsDir).mkdirs();
        
        String filename = resultsDir + "/benchmark_results_java";
        if (methodSuffix != null && !methodSuffix.isEmpty()) {
            filename += "_" + methodSuffix;
        }
        if (clusterSize > 0) {
            filename += "_" + clusterSize + "nodes";
        }
        filename += "_" + System.currentTimeMillis() + ".csv";
        
        try (FileWriter writer = new FileWriter(filename)) {
            // Write header
            writer.write("matrix_size,method,run,time_ms,network_time_ms,data_transferred_kb,memory_kb,cluster_nodes,memory_per_node_kb\n");
            
            // Write data
            for (BenchmarkResult result : results) {
                // Basic times
                for (int i = 0; i < result.basicTimes.length; i++) {
                    writer.write(result.size + ",basic," + (i+1) + "," + String.format(Locale.US, "%.2f", result.basicTimes[i]) + ",0,0," + 
                                String.format(Locale.US, "%.2f", result.memoryUsed[i]) + ",0,0\n");
                }
                
                // Parallel times
                for (int i = 0; i < result.parallelTimes.length; i++) {
                    writer.write(result.size + ",parallel," + (i+1) + "," + String.format(Locale.US, "%.2f", result.parallelTimes[i]) + ",0,0," + 
                                String.format(Locale.US, "%.2f", result.parallelMemory[i]) + ",0,0\n");
                }
                
                // Distributed times
                if (result.distributedTimes != null) {
                    for (int i = 0; i < result.distributedTimes.length; i++) {
                        writer.write(result.size + ",distributed," + (i+1) + "," + String.format(Locale.US, "%.2f", result.distributedTimes[i]) + "," +
                                    String.format(Locale.US, "%.2f", result.networkTimes[i]) + "," + String.format(Locale.US, "%.2f", result.dataTransferred[i] * 1024) + "," + 
                                    String.format(Locale.US, "%.2f", result.distributedMemory[i]) + "," + result.clusterNodes + "," +
                                    String.format(Locale.US, "%.2f", result.memoryPerNodeKB) + "\n");
                    }
                }
            }
            
            System.out.println("\nResults saved to: " + filename);
            
        } catch (IOException e) {
            System.err.println("Error saving results: " + e.getMessage());
        }
    }
    
    // Helper methods
    private static double average(long[] values) {
        long sum = 0;
        for (long value : values) {
            sum += value;
        }
        return (double) sum / values.length;
    }
    
    private static long min(long[] values) {
        long min = values[0];
        for (long value : values) {
            if (value < min) min = value;
        }
        return min;
    }
    
    private static long max(long[] values) {
        long max = values[0];
        for (long value : values) {
            if (value > max) max = value;
        }
        return max;
    }
    
    private static String arrayToString(int[] arr) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < arr.length; i++) {
            sb.append(arr[i]);
            if (i < arr.length - 1) sb.append(", ");
        }
        return sb.toString();
    }
    
    private static double averageDouble(double[] values) {
        double sum = 0;
        for (double value : values) {
            sum += value;
        }
        return sum / values.length;
    }
    
    private static double minDouble(double[] values) {
        double min = values[0];
        for (double value : values) {
            if (value < min) min = value;
        }
        return min;
    }
    
    private static double maxDouble(double[] values) {
        double max = values[0];
        for (double value : values) {
            if (value > max) max = value;
        }
        return max;
    }
    
    private static double[] convertToDouble(long[] values) {
        double[] result = new double[values.length];
        for (int i = 0; i < values.length; i++) {
            result[i] = (double) values[i];
        }
        return result;
    }
    
    // Inner class to store results
    private static class BenchmarkResult {
        int size;
        double[] basicTimes;
        double[] parallelTimes;
        double[] distributedTimes;
        double[] networkTimes;
        double[] dataTransferred;
        double[] memoryUsed;          // For basic (KB with decimals)
        double[] parallelMemory;      // For parallel (KB with decimals)
        double[] distributedMemory;   // For distributed (KB with decimals)
        int clusterNodes;
        double memoryPerNodeKB;     // Estimated memory per cluster node in KB
        
        BenchmarkResult(int size) {
            this.size = size;
        }
    }
}